src/bit-systems/link-hover-menu.ts

```diff
@@ -1,47 +1,150 @@
+    return;
+  }
+
+  const hovered = hoveredQuery(world);
+  const target = hovered.map(eid => findAncestorWithComponent(world, Link, eid))[0] || 0;
+  if (target) {
+    LinkHoverMenu.targetObjectRef[menu] = target;
+    LinkHoverMenu.clearTargetTimer[menu] = world.time.elapsed + 1000;
+    return;
+  }
+
+  if (hovered.some(eid => findAncestorWithComponent(world, LinkHoverMenu, eid))) {
+    LinkHoverMenu.clearTargetTimer[menu] = world.time.elapsed + 1000;
+    return;
+  }
+
+  if (world.time.elapsed > LinkHoverMenu.clearTargetTimer[menu]) {
+    LinkHoverMenu.targetObjectRef[menu] = 0;
+    return;
+  }
+}
+
+async function handleLinkClick(world: HubsWorld, button: EntityID) {
+  const exitImmersive = async () => await handleExitTo2DInterstitial(false, () => {}, true);
+
+  const menu = findAncestorWithComponent(world, LinkHoverMenu, button)!;
+  const linkEid = LinkHoverMenu.targetObjectRef[menu];
+  const src = APP.getString(Link.url[linkEid])!;
+  const url = new URL(src);
+  const linkType = Link.type[linkEid];
+  switch (linkType) {
+    case LinkType.LINK:
+      await exitImmersive();
+      window.open(src);
+      break;
+    case LinkType.AVATAR:
+      const avatarId = new URL(src).pathname.split("/").pop();
+      window.APP.store.update({ profile: { avatarId } });
+      APP.scene!.emit("avatar_updated");
+      break;
+    case LinkType.SCENE:
+      APP.scene!.emit("scene_media_selected", src);
+      break;
+    case LinkType.WAYPOINT:
+      // move to waypoint w/o writing to history
+      window.history.replaceState(null, "", window.location.href.split("#")[0] + url.hash);
+      break;
+    case LinkType.ROOM:
+      const waypoint = url.hash && url.hash.substring(1);
+      // move to new room without page load or entry flow
+      const hubId = hubIdFromUrl(url);
+      changeHub(hubId, true, waypoint);
+      break;
+    case LinkType.ROOM_URL:
+      await exitImmersive();
+      location.href = src;
+      break;
+  }
+}
+
+function moveToTarget(world: HubsWorld, menu: EntityID) {
+  const linkEid = LinkHoverMenu.targetObjectRef[menu];
+  const targetObject = world.eid2obj.get(linkEid)!;
+  targetObject.updateMatrices();
+  const menuObject = world.eid2obj.get(menu)!;
+  setMatrixWorld(menuObject, targetObject.matrixWorld);
+}
+
+function updateButtonText(world: HubsWorld, menu: EntityID, button: EntityID) {
+  const text = findChildWithComponent(world, TextTag, button)!;
+  const textObj = world.eid2obj.get(text)! as TroikaText;
+  const linkEid = LinkHoverMenu.targetObjectRef[menu];
+  const linkType = Link.type[linkEid];
+  let label = "open link";
+  switch (linkType) {
+    case LinkType.AVATAR:
+      label = "use avatar";
+      break;
+    case LinkType.SCENE:
+      label = "use scene";
+      break;
+    case LinkType.WAYPOINT:
+      label = "go to";
+      break;
+    case LinkType.ROOM:
+    case LinkType.ROOM_URL:
+      label = "visit room";
+      break;
+  }
```

Comment:
I guess this is mostly because `textObj.text` doesn't accept undefined and if you not assign upfront the compiler complains because it can be undefined. We can initialize as empty and keep the case if that makes it clearer.

Reply:
I think initializing with empty string is ok, thanks.